src/main/java/javaslang/collection/HashSet.java

```diff
@@ -350,7 +360,7 @@ public boolean isEmpty() {
 
     @Override
     public HashSet<T> remove(T element) {
-        return new HashSet<>(tree.remove(element));
+        return contains(element) ? new HashSet<>(tree.remove(element)) : this;
```

Comment:
what about the tree? does it return the same instance when removing an element which is not contained? if so, we may compare tree == removed, where removed = tree.remove(element)